File: web/src/cross-pollination.ts
import { selectAuthor } from "./investigation";
import { escapeHtml } from "./post-display";
import type {
  CrossPollinationActor,
  CrossPollinationNarrativeRef,
  CrossPollinationReport,
  NarrativePollinationHits,
  NarrativePairOverlap,
} from "./types";

/** Distinct hues for narrative silos — stable by narrative_id. */
const NARRATIVE_PALETTE = [
  { bg: "#2980b9", fg: "#ecf6ff", border: "#5dade2" },
  { bg: "#c0392b", fg: "#fff0ee", border: "#e74c3c" },
  { bg: "#27ae60", fg: "#eafff3", border: "#58d68d" },
  { bg: "#8e44ad", fg: "#f9eeff", border: "#bb8fce" },
  { bg: "#d68910", fg: "#fff8e8", border: "#f4d03f" },
  { bg: "#16a085", fg: "#e8fff9", border: "#48c9b0" },
  { bg: "#c2185b", fg: "#ffeef4", border: "#f06292" },
  { bg: "#5c6bc0", fg: "#eef0ff", border: "#9fa8da" },
];

function narrativePaletteIndex(narrativeId: number): number {
  return Math.abs(narrativeId) % NARRATIVE_PALETTE.length;
}

function shortenNarrativeName(name: string, max = 18): string {
  const t = name.trim();
  if (t.length <= max) return t;
  return `${t.slice(0, max - 1)}…`;
}

function collectNarrativesFromReport(report: CrossPollinationReport | null): CrossPollinationNarrativeRef[] {
  const byId = new Map<number, CrossPollinationNarrativeRef>();
  for (const actor of report?.actors ?? []) {
    for (const n of actor.narratives) {
      if (!byId.has(n.narrative_id)) byId.set(n.narrative_id, n);
    }
  }
  return [...byId.values()].sort((a, b) => a.narrative_name.localeCompare(b.narrative_name));
}

function renderNarrativePill(
  ref: CrossPollinationNarrativeRef,
  options?: { current?: boolean; also?: boolean }
): string {
  const idx = narrativePaletteIndex(ref.narrative_id);
  const short = shortenNarrativeName(ref.narrative_name);
  const title = `${ref.narrative_name} · ${ref.post_count} post(s)`;
  const flags = [
    options?.current ? "cross-narr-pill-current" : "",
    options?.also ? "cross-narr-pill-also" : "",
  ]
    .filter(Boolean)
    .join(" ");
  return `<span
    class="cross-narr-pill cross-narr-pill-${idx}${flags ? ` ${flags}` : ""}"
    style="--pill-bg:${NARRATIVE_PALETTE[idx]!.bg};--pill-fg:${NARRATIVE_PALETTE[idx]!.fg};--pill-border:${NARRATIVE_PALETTE[idx]!.border}"
    title="${escapeHtml(title)}"
  >${escapeHtml(short)}</span>`;
}

function renderIntersectionBadges(
  narratives: CrossPollinationNarrativeRef[],
  options?: { currentNarrativeId?: number; onlyOther?: boolean }
): string {
  if (!narratives.length) return "";
  const sorted = [...narratives].sort((a, b) => a.narrative_name.localeCompare(b.narrative_name));
  const pills = sorted
    .filter((n) => {
      if (!options?.onlyOther || options.currentNarrativeId == null) return true;
      return n.narrative_id !== options.currentNarrativeId;
    })
    .map((n) =>
      renderNarrativePill(n, {
        current: options?.currentNarrativeId === n.narrative_id,
        also:
          options?.currentNarrativeId != null &&
          n.narrative_id !== options.currentNarrativeId,
      })
    )
    .join("");

  if (!pills) return "";

  const label = sorted.map((n) => n.narrative_name).join(", ");
  return `<span class="cross-narrative-pills" aria-label="Narrative overlap: ${escapeHtml(label)}">${pills}</span>`;
}

function renderNarrativeLegend(report: CrossPollinationReport): string {
  const narratives = collectNarrativesFromReport(report);
  if (narratives.length < 2) return "";
  const items = narratives
    .map((n) => renderNarrativePill(n))
    .join('<span class="cross-legend-sep" aria-hidden="true">·</span>');
  return `<div class="cross-narrative-legend" role="list" aria-label="Narrative color key">${items}</div>`;
}

function renderPairRow(pair: NarrativePairOverlap): string {
  const a: CrossPollinationNarrativeRef = {
    narrative_id: pair.narrative_a_id,
    narrative_name: pair.narrative_a_name,
    post_count: 0,
    max_outrage: null,
    first_seen: null,
    last_seen: null,
  };
  const b: CrossPollinationNarrativeRef = {
    narrative_id: pair.narrative_b_id,
    narrative_name: pair.narrative_b_name,
    post_count: 0,
    max_outrage: null,
    first_seen: null,
    last_seen: null,
  };
  return `<li class="cross-pair-item">
    <span class="cross-pair-badges">${renderNarrativePill(a)}<span class="cross-pair-link" aria-hidden="true">×</span>${renderNarrativePill(b)}</span>
    <span class="cross-pair-count">${pair.shared_actor_count} shared</span>
  </li>`;
}

export function crossPollinationPanelHtml(
  report: CrossPollinationReport | null,
  asInner = false
): string {
  const count = report?.actor_count ?? 0;
  const badge =
    count > 0
      ? `<span class="topology-badge topology-star">${count} cross-narrative actors</span>`
      : `<span class="topology-badge topology-sparse">global scan</span>`;

  const inner = `
      <h2>Narrative cross-pollination ${badge}</h2>
      <p class="chart-caption">
        Scans <strong>all narratives</strong> in heimdall.db for accounts posting in multiple keyword silos.
        <strong>Colored pills</strong> = which campaigns share each actor (hover for post counts).
      </p>
      <div id="cross-pollination-global-host"></div>
      <div id="cross-pollination-narrative-host"></div>
  `;
  if (asInner) {
    return `<div class="panel panel-cross-pollination" id="cross-pollination-panel">${inner}</div>`;
  }
  return `<section class="panel panel-cross-pollination" id="cross-pollination-panel">${inner}</section>`;
}

export function renderGlobalCrossPollination(
  host: HTMLElement,
  report: CrossPollinationReport | null
): void {
  if (!report?.available) {
    host.innerHTML = "<p class='empty'>Cross-pollination scan unavailable (no posts in database).</p>";
    return;
  }
  if (!report.actor_count) {
    host.innerHTML =
      "<p class='empty'>No actors span multiple narratives yet — ingest at least two narratives with shared accounts.</p>";
    return;
  }

  const legend = renderNarrativeLegend(report);
  const pairs = (report.narrative_pairs ?? []).slice(0, 6).map(renderPairRow).join("");

  const actors = (report.actors ?? [])
    .slice(0, 12)
    .map((a) => renderActorButton(a))
    .join("");

  host.innerHTML = `
    ${legend}
    <h3 class="cross-subhead">Top cross-narrative actors</h3>
    <div class="cross-actor-list">${actors}</div>
    ${
      pairs
        ? `<h3 class="cross-subhead">Narrative pair overlap</h3><ul class="cross-pair-list">${pairs}</ul>`
        : ""
    }
  `;

  bindActorButtons(host);
}

export function renderNarrativeCrossPollination(
  host: HTMLElement,
  hits: NarrativePollinationHits | null,
  narrativeName: string,
  currentNarrativeId?: number
): void {
  if (!hits || hits.hit_count === 0) {
    host.innerHTML = `<p class="metric-sub">No multi-narrative actors in <strong>${escapeHtml(narrativeName)}</strong> yet.</p>`;
    return;
  }

  host.innerHTML = `
    <h3 class="cross-subhead">In this narrative (${hits.hit_count} multi-silo actors)</h3>
    <p class="cross-caption-inline"><span class="cross-narr-pill-legend-hint">Pill with ring</span> = this narrative · other colors = also active elsewhere</p>
    <div class="cross-actor-list">${hits.actors
      .slice(0, 8)
      .map((a) => renderActorButton(a, { currentNarrativeId }))
      .join("")}</div>
  `;

  bindActorButtons(host);
}

function renderActorButton(
  actor: CrossPollinationActor,
  options?: { currentNarrativeId?: number }
): string {
  const label = actor.author_handle ?? actor.author_id.slice(0, 14);
  const score = actor.pollination_score.toFixed(2);

  let narratives = actor.narratives;
  if (actor.other_narratives?.length) {
    const currentId = options?.currentNarrativeId ?? hitsCurrentFromActor(actor);
    const current = narratives.find((n) => n.narrative_id === currentId);
    narratives = current
      ? [current, ...actor.other_narratives]
      : [...actor.other_narratives];
  }

  const pills = renderIntersectionBadges(narratives, {
    currentNarrativeId: options?.currentNarrativeId,
  });

  return `<button type="button" class="cross-actor-btn" data-author-id="${escapeHtml(actor.author_id)}" data-author-label="${escapeHtml(label)}">
    <div class="cross-actor-head">
      <strong class="cross-actor-handle">${escapeHtml(label)}</strong>
      ${pills}
    </div>
    <span class="cross-actor-meta">${actor.narrative_count} narratives · ${actor.total_posts} posts · score ${score}</span>
  </button>`;
}

function hitsCurrentFromActor(actor: CrossPollinationActor): number | undefined {
  if (!actor.other_narratives?.length) return undefined;
  const otherIds = new Set(actor.other_narratives.map((n) => n.narrative_id));
  const current = actor.narratives.find((n) => !otherIds.has(n.narrative_id));
  return current?.narrative_id;
}

function bindActorButtons(root: ParentNode): void {
  root.querySelectorAll<HTMLButtonElement>(".cross-actor-btn").forEach((btn) => {
    btn.addEventListener("click", () => {
      const id = btn.dataset.authorId;
      const label = btn.dataset.authorLabel ?? id;
      if (id) selectAuthor(id, `Cross-poll: ${label}`);
      window.dispatchEvent(new CustomEvent("heimdall:goto-posts"));
    });
  });
}

File: heimdall/analysis/cross_pollination.py
"""Cross-narrative author overlap — threat actors spanning multiple keyword silos."""

from __future__ import annotations

import math
from dataclasses import dataclass
from datetime import datetime


@dataclass(frozen=True)
class NarrativePresence:
    narrative_id: int
    narrative_name: str
    post_count: int
    max_outrage: float | None
    first_seen: str | None
    last_seen: str | None


@dataclass(frozen=True)
class CrossPollinationActor:
    """Author account active in multiple narratives (organized proxy indicator)."""

    actor_key: str
    platform: str
    author_id: str
    author_handle: str | None
    narrative_count: int
    total_posts: int
    pollination_score: float
    narratives: list[NarrativePresence]
    span_days: float


@dataclass(frozen=True)
class NarrativePairOverlap:
    narrative_a_id: int
    narrative_a_name: str
    narrative_b_id: int
    narrative_b_name: str
    shared_actor_count: int


def _actor_key(platform: str, author_id: str) -> str:
    return f"{(platform or 'unknown').lower()}:{author_id}"


def _parse_iso(value: str | datetime | None) -> datetime | None:
    if value is None:
        return None
    if isinstance(value, datetime):
        return value
    raw = str(value).strip()
    if not raw:
        return None
    try:
        return datetime.fromisoformat(raw.replace("Z", "+00:00"))
    except ValueError:
        return None


def _span_days(timestamps: list[datetime | None]) -> float:
    valid = [t for t in timestamps if t is not None]
    if len(valid) < 2:
        return 0.0
    return round((max(valid) - min(valid)).total_seconds() / 86400, 2)


def pollination_score(narrative_count: int, total_posts: int, span_days: float) -> float:
    """
    Higher when an actor spans more narratives with sustained volume.

    Roughly: narratives * log2(posts+1) * mild span bonus (capped).
    """
    if narrative_count < 2:
        return 0.0
    base = narrative_count * math.log2(total_posts + 1)
    span_bonus = min(1.5, 1.0 + span_days / 30.0) if span_days > 0 else 1.0
    return round(base * span_bonus, 4)


def build_cross_pollination_report(
    rows: list[tuple],
    *,
    min_narratives: int = 2,
    min_posts_total: int = 2,
    max_actors: int = 50,
    max_pairs: int = 30,
) -> dict:
    """
    Build cross-pollination report from aggregated DB rows.

    Each row:
      (narrative_id, narrative_name, platform, author_id, author_handle,
       post_count, max_outrage, first_seen, last_seen)
    """
    by_actor: dict[str, dict] = {}

    for row in rows:
        (
            narrative_id,
            narrative_name,
            platform,
            author_id,
            author_handle,
            post_count,
            max_outrage,
            first_seen,
            last_seen,
        ) = row
        if not author_id:
            continue
        key = _actor_key(str(platform), str(author_id))
        bucket = by_actor.setdefault(
            key,
            {
                "platform": str(platform),
                "author_id": str(author_id),
                "author_handle": None,
                "narratives": [],
                "timestamps": [],
            },
        )
        if author_handle and not bucket["author_handle"]:
            bucket["author_handle"] = str(author_handle)
        bucket["narratives"].append(
            NarrativePresence(
                narrative_id=int(narrative_id),
                narrative_name=str(narrative_name),
                post_count=int(post_count),
                max_outrage=float(max_outrage) if max_outrage is not None else None,
                first_seen=str(first_seen) if first_seen else None,
                last_seen=str(last_seen) if last_seen else None,
            )
        )
        for ts in (first_seen, last_seen):
            parsed = _parse_iso(ts)
            if parsed:
                bucket["timestamps"].append(parsed)

    actors: list[CrossPollinationActor] = []
    for key, bucket in by_actor.items():
        narratives = sorted(bucket["narratives"], key=lambda n: (-n.post_count, n.narrative_name))
        narrative_count = len(narratives)
        total_posts = sum(n.post_count for n in narratives)
        if narrative_count < min_narratives or total_posts < min_posts_total:
            continue
        span = _span_days(bucket["timestamps"])
        score = pollination_score(narrative_count, total_posts, span)
        actors.append(
            CrossPollinationActor(
                actor_key=key,
                platform=bucket["platform"],
                author_id=bucket["author_id"],
                author_handle=bucket["author_handle"],
                narrative_count=narrative_count,
                total_posts=total_posts,
                pollination_score=score,
                narratives=narratives,
                span_days=span,
            )
        )

    actors.sort(key=lambda a: (-a.pollination_score, -a.narrative_count, -a.total_posts))
    actors = actors[:max_actors]

    # Narrative pair overlap (shared actors)
    narrative_ids = sorted({n.narrative_id for a in actors for n in a.narratives})
    id_to_name = {n.narrative_id: n.narrative_name for a in actors for n in a.narratives}
    pair_counts: dict[tuple[int, int], int] = {}

    for actor in actors:
        nids = sorted({n.narrative_id for n in actor.narratives})
        for i in range(len(nids)):
            for j in range(i + 1, len(nids)):
                pair = (nids[i], nids[j])
                pair_counts[pair] = pair_counts.get(pair, 0) + 1

    pairs: list[NarrativePairOverlap] = []
    for (a_id, b_id), count in sorted(pair_counts.items(), key=lambda x: -x[1]):
        pairs.append(
            NarrativePairOverlap(
                narrative_a_id=a_id,
                narrative_a_name=id_to_name.get(a_id, str(a_id)),
                narrative_b_id=b_id,
                narrative_b_name=id_to_name.get(b_id, str(b_id)),
                shared_actor_count=count,
            )
        )
    pairs = pairs[:max_pairs]

    return {
        "min_narratives": min_narratives,
        "actor_count": len(actors),
        "narrative_count": len(narrative_ids),
        "actors": [_actor_to_dict(a) for a in actors],
        "narrative_pairs": [_pair_to_dict(p) for p in pairs],
    }


def _actor_to_dict(actor: CrossPollinationActor) -> dict:
    return {
        "actor_key": actor.actor_key,
        "platform": actor.platform,
        "author_id": actor.author_id,
        "author_handle": actor.author_handle,
        "narrative_count": actor.narrative_count,
        "total_posts": actor.total_posts,
        "pollination_score": actor.pollination_score,
        "span_days": actor.span_days,
        "narratives": [
            {
                "narrative_id": n.narrative_id,
                "narrative_name": n.narrative_name,
                "post_count": n.post_count,
                "max_outrage": n.max_outrage,
                "first_seen": n.first_seen,
                "last_seen": n.last_seen,
            }
            for n in actor.narratives
        ],
    }


def _pair_to_dict(pair: NarrativePairOverlap) -> dict:
    return {
        "narrative_a_id": pair.narrative_a_id,
        "narrative_a_name": pair.narrative_a_name,
        "narrative_b_id": pair.narrative_b_id,
        "narrative_b_name": pair.narrative_b_name,
        "shared_actor_count": pair.shared_actor_count,
    }


def narrative_cross_pollination_hits(
    report: dict,
    narrative_id: int,
    *,
    max_hits: int = 25,
) -> dict:
    """Actors in this narrative who also appear in other narratives."""
    hits: list[dict] = []
    for actor in report.get("actors", []):
        narratives = actor.get("narratives", [])
        if not any(n.get("narrative_id") == narrative_id for n in narratives):
            continue
        other = [n for n in narratives if n.get("narrative_id") != narrative_id]
        if not other:
            continue
        hits.append(
            {
                **actor,
                "other_narratives": other,
                "other_narrative_count": len(other),
            }
        )
    hits.sort(key=lambda h: (-h.get("pollination_score", 0), -h.get("other_narrative_count", 0)))
    return {
        "narrative_id": narrative_id,
        "hit_count": len(hits),
        "actors": hits[:max_hits],
    }


def cross_pollination_cib_signals(report: dict, narrative_id: int) -> list[str]:
    """Signals when a narrative hosts multi-narrative actors."""
    hits = narrative_cross_pollination_hits(report, narrative_id)
    count = hits.get("hit_count", 0)
    if count == 0:
        return []
    top = hits.get("actors", [])[:3]
    signals = [f"cross_pollination_{count}_multi_narrative_actors"]
    for actor in top:
        names = ", ".join(
            n.get("narrative_name", "?") for n in actor.get("other_narratives", [])[:3]
        )
        handle = actor.get("author_handle") or actor.get("author_id", "")[:12]
        signals.append(f"cross_pollination_actor_{handle}_also_in_{names}")
    return signals
